Bluetooth: Mesh: Generic OnOff, Generic Level, Lighting & Vendor Models demo

vikrant8052 opened the pull request:
This is a application demonstrating a Bluetooth mesh node in
which Root element has following models

- Generic OnOff Server
- Generic OnOff Client
- Generic Level Server
- Generic Level Client
- Generic Power OnOff Server
- Generic Power OnOff Setup Server
- Generic Power OnOff Client
- Light Lightness Server
- Light Lightness Setup Server
- Light Lightness Client
- Light CTL Server
- Light CTL Setup Server
- Light CTL Client
- Vendor Model

And Secondary element has following models

- Generic OnOff Server
- Generic OnOff Client
- Generic Level Server
- Generic Level Client
- Light CTL Temperature Server


Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg @dbkinder 
Please review it. All checks have passed, so now it will be easy for us to go ahead.

Thank You !!

dbkinder reviewed (approved):
+1 for docs

vikrant8051 commented:
Thank you @dbkinder.

vikrant8051 commented:
@jhedberg 
Please review it. Your feedback is most important.

vikrant8051 commented:
@jhedberg @carlescufi 
Please review it. Waiting for your reply.
Sorry for inconvenience caused due to removing old PR.

vikrant8051 commented:
So '\n' are removing

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+	       src/common.c
+	       src/mesh/ble_mesh.c
+               src/mesh/device_composition.c
+               src/mesh/publisher.c
```
Minor, but looks like the above two lines have different indentation than the others.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/common.c:
```diff
@@ -0,0 +1,129 @@
+{
+	if (nvs_write(&fs, NVS_LED_STATE_ID, &light_state_current,
+	    sizeof(struct light_state_t)) > 0) {
+
```
Minor: remove this redundant empty line

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/common.c:
```diff
@@ -0,0 +1,129 @@
+	    sizeof(struct light_state_t)) > 0) {
+
+		printk("Light state has saved !!\n");
+
```
same here

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/common.c:
```diff
@@ -0,0 +1,129 @@
+		printk("Light state has saved !!\n");
+
+	} else {
+
```
and here

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/common.h:
```diff
@@ -0,0 +1,85 @@
+#define LEVEL_100 32767
+
+struct light_state_t {
+
```
redundant line

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/common.h:
```diff
@@ -0,0 +1,85 @@
+
+struct light_state_t {
+
+	u8_t OnOff;
```
Why CamelCase here? To be consistent with the rest of the app (and the Zephyr source tree in general) I'd name this to on_off.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/main.c:
```diff
@@ -0,0 +1,38 @@
+
+void main(void)
+{
+	int err, rc = 0;
```
Unnecessary initialization to zero upon declaration

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/main.c:
```diff
@@ -0,0 +1,38 @@
+	int err, rc = 0;
+
+	rc = nvs_init(&fs, FLASH_DEV_NAME, STORAGE_MAGIC);
+
```
redundant empty line (generally, no empty line between assignment to variable and testing for the variable's value)

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/main.c:
```diff
@@ -0,0 +1,38 @@
+	rc = nvs_init(&fs, FLASH_DEV_NAME, STORAGE_MAGIC);
+
+	if (rc) {
+		printk("Flash Init failed\n");
```
missing return statement? If nvs_init failed I don't think it's save to do e.g. an nvs_read() call.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/ble_mesh.c:
```diff
@@ -0,0 +1,85 @@
+	/* board_init(); */
+
+	err = bt_mesh_init(&prov, &comp);
+
```
redundant empty line (no empty line between assignment to variable and testing its value)

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+			  struct bt_mesh_msg_ctx *ctx,
+			  struct net_buf_simple *buf)
+{
+	int err = 0;
```
I'd move this variable inside the smallest scope where it's used, i.e. to the beginning of the `if (model->pub->addr != BT_MESH_ADDR_UNASSIGNED)` branch. Also, there's no need to initialize it to zero.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+		net_buf_simple_add_u8(msg, state->current);
+
+		err = bt_mesh_model_publish(model);
+
```
redundant empty line

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+			  struct bt_mesh_msg_ctx *ctx,
+			  struct net_buf_simple *buf)
+{
+	int err;
```
move this inside the smaller scope where it's used (further down this function

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+	if (model->pub->addr != BT_MESH_ADDR_UNASSIGNED) {
+
+		bt_mesh_model_msg_init(msg,
+				       BT_MESH_MODEL_OP_2(0x82, 0x08));
```
Would be good to have a define for what message (0x82, 0x08) is

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+		net_buf_simple_add_le16(msg, state->current);
+
+		err = bt_mesh_model_publish(model);
+
```
redundant empty line

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+	int err;
+	struct net_buf_simple *msg = model->pub->msg;
+	struct server_state *state = model->user_data;
+
```
Redundant empty line. I'd prefer to have any initializations that happen from return values of function calls to be done separately from the variable declarations (concerning the two lines below).

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+
+	if (tmp32 < -32768) {
+		tmp32 = -32768;
+
```
redundant empty line

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+		tmp32 = 32767;
+	}
+
+	state->current = (int16_t)tmp32;
```
Is this typecast really needed?

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+	}
+
+	if (model->pub->addr != BT_MESH_ADDR_UNASSIGNED) {
+
```
redundant empty line

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+	if (model->pub->addr != BT_MESH_ADDR_UNASSIGNED) {
+
+		bt_mesh_model_msg_init(msg,
+				       BT_MESH_MODEL_OP_2(0x82, 0x08));
```
Please add a define for what (0x82, 0x08) is. This sample is supposed to serve as a good reference for developers, so it must be easily readable and understandable.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+		net_buf_simple_add_le16(msg, state->current);
+
+		err = bt_mesh_model_publish(model);
+
```
redundant empty line

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+				 struct bt_mesh_msg_ctx *ctx,
+				 struct net_buf_simple *buf)
+{
+	s16_t tmp16 = net_buf_simple_pull_le16(buf);
```
There should always be an empty line after variable declarations, however in this case you could just remove the variable and do the call to net_buf_simple_pull_le16() from the parameters of the printk() call.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+};
+
+struct bt_mesh_model root_models[] = {
+
```
redundant empty line

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_app/src/mesh/device_composition.c:
```diff
@@ -0,0 +1,340 @@
+};
+
+static struct bt_mesh_model s0_models[] = {
+
```
redundant empty line, you're also inconsistent with your own style since in some array declarations you don't have this empty line (e.g. elements array below)

jhedberg reviewed (changes requested):
I hope you have patience for one more review round :)

These are mostly minor coding style issues, but there's at least one thing I consider a bug: doing nvs calls even if nvs_init fails.

Also, where's your PR for the existing onoff_app's settings support? You had those changes in a commit already (but in the wrong PR), but now I can't find that anymore?

jhedberg commented:
One other thing: please squash your commits together. The "fixup" commits would only be justified if the initial app was already merged.

vikrant8051 commented:
I've make suggested changes.

I'll send separate PR for onoff-app. Previously, I tried but it get merge with this.

Handling git & Github related commands is only problem for me
since I don't have any experience. For removing bugs & redundancies I'm 
always ready.

vikrant8051 commented:
@jhedberg 
I've make suggested changes.

I'll send separate PR for onoff-app. Previously, I tried but it get merge with this.

Handling git & Github related commands is only problem for me
since I don't have any experience. For removing bugs & redundancies I'm 
always ready.

carlescufi commented:
>Handling git & Github related commands is only problem for me
since I don't have any experience.

@vikrant8051 it really is quite easy.

You just need to add a remote to your local repository like this:
`git remote add upstream git@github.com:zephyrproject-rtos/zephyr.git`
then you can do this:
```
git fetch upstream
git rebase -i upstream/master
```
and then you can select which commits to modify or to squash together.
There is more info here:
http://docs.zephyrproject.org/contribute/contribute_guidelines.html#contribution-workflow

vikrant8051 commented:
I've done it !!

Delete local directory. Since I had directly cloned my branch. 
Download "my" master branch & cd into it &

Execute -> 
1) git reset origin/master
2) git checkout -b mesh_demo
3) add "onoff_level_app" in it
4) git add -A
5) git commit -S -s (edit COMMIT_EDITMSG file)
6) git push -f origin mesh_demo

Hi @carlescufi, thanks for your help. I'll follow your steps next times, if required :)

vikrant8051 commented:
Hi @sjanc,
Please review this PR.

carlescufi commented on samples/boards/nrf52/mesh/onoff_level_app/src/common.c:
```diff
@@ -0,0 +1,128 @@
+{
+	static struct gpio_callback button_cb[4];
+
+	/* --------------- LEDs configiuratin & setting --------------- */
```
Can you get rid of these `/* ----- ----*/` lines? I don't think we use them elsewhere in Zephyr, so trying to keep the code consistent

vikrant8051 commented:
Hi @jhedberg & @carlescufi,
I've added support for Vendor Model too. Please review & approved it if everything is OK.

vikrant8051 commented:
@jhedberg @carlescufi 

My concern is only regarding this definition,

/* Start address of the filesystem in flash */
#define NVS_STORAGE_OFFSET (FLASH_AREA_STORAGE_OFFSET - 8192)

which is defined in common.h.

sjanc reviewed (commented):
The code looks OK to me but all source files are missing license headers.

vikrant8051 commented:
Hi @sjanc , thanks for review. I've make suggested changes.

vikrant8051 commented:
@jhedberg 
Now I've completely removed #NVS & its dependent code from this PR.

And enable following options 
CONFIG_BT_RX_STACK_SIZE=4096
CONFIG_BT_MESH_RPL_STORE_TIMEOUT=5

Now do following steps:
1) Provision DEVICE using #nRFMesh 
2) Configure available models
3) Play with button & GUI for testing
4) Unprovisioned NODE using node-reset command
5) Re-provision DEVICE using #nRFMesh 
6) Configure available models
7) Play with button & GUI for testing
8) Reset the Board.

Now here I found device in Provisioned state, but Buttons & LED were not at all working.
So I reconnect & try to control LEDs using APP GUI buttons but they didn't work.
So reconfigure using #nRFMesh. Then it worked as expected.

I don't know where is problem. But this is my observations. Please do testing with this App. 

Thank You !!

vikrant8051 commented:
@dbkinder I've added Vendor Model support. So edited README.rst as per latest changes.
Please review it again.

vikrant8051 commented:
@jhedberg 

Hi Johan,

I've added following things in existing samples/boards/nrf52/mesh/onoff-app

1) #include <settings/settings.h> in main.c

2)
    if (IS_ENABLED(CONFIG_SETTINGS)) {
        settings_load();
    } 

    above lines in bt_ready()

3)

CONFIG_BT_SETTINGS=y
CONFIG_FLASH=y
CONFIG_FLASH_PAGE_LAYOUT=y
CONFIG_FLASH_MAP=y
CONFIG_FCB=y
CONFIG_SETTINGS=y
CONFIG_SETTINGS_FCB=y

 above lines in prj.conf

And do following steps (only configured GEN_Servers)... 

Provisioned -> configured -> test LED control using #nRFMesh GUI button -> unprovisioned -> re-provisioned ->
.. re- configured -> test LED control using #nRFMesh GUI button -> Reset the board  -> found device in Provisioned state
..-> make connection with #nRFMesh & proxy node -> try to control LEDs using APP GUI Button -> And there is no effect

I've to re-bind APP key with Models.

This is same observation which happened for this PR's app.

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------

vikrant8051 commented:
@jhedberg 
After sometime I again reset the board. Now I found device in unprovisioned state !!
( this is happening in case of samples/boards/nrf52/mesh/onoff-app)

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_vnd_app/src/common.h:
```diff
@@ -0,0 +1,67 @@
+
+#include "ble_mesh.h"
+#include "device_composition.h"
+#include "publisher.h"
```
Please move all the includes to the c-files that need them. This kind of indirect inclusion is not really used in the tree. The only exception is public header files and even then the only includes they have is for things that's part of the header file itself (here neither one of these is true).

jhedberg reviewed (changes requested):
Sorry for the delay in doing another review. Two small things that still need fixing

 - The includes in common.h (see my more detailed comment)

 - Remove the merge commit (rebasing against upstream master should take care of it)

vikrant8051 commented:
@jhedberg 
I've worked suggested changes. 

But what about my observations ?
Perisistent data storage is not reliable if you unprovisioned & re-provision device.

I don't know from where that 1st commit come. I will work to remove it once you OK with everything.

vikrant8051 commented:
@jhedberg 
What should be ideal value of 
CONFIG_BT_MESH_RPL_STORE_TIMEOUT ?

I our previous communication you recommended it as 5 but in samples/bluetooth/mesh_demo app you have set it to 600.

vikrant8051 commented:
@jhedberg Removed the merge commit.

vikrant8051 commented:
@jhedberg 
Do not merge it. I will try to merge Light Lightness (0x1300)  & Light CTL (0x1303) Server Model too.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_vnd_app/src/common.c:
```diff
@@ -0,0 +1,124 @@
+/*  Bluetooth: Mesh Generic OnOff, Generic Level & Vendor Models
+ *
+ * Copyright (c) 2018 Vikrant
```
Could you please use your full name here, so that it matches the git author and signed-off-by. Also, are you sure that the copyright belongs to you and not your company? Most of the time it's the employer that owns the work-related code that's written by an employee.

jhedberg replied:
One other thing: it'd be appropriate to also have a second copyright line from the onoff-app for any files where you've derived code from that app.

jhedberg replied:
Actually, just forget about this for now - I just realized that the author (Steve Brown) didn't put the right copyright to begin with. So just leave it with your own line.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_vnd_app/src/common.h:
```diff
@@ -0,0 +1,62 @@
+#include <stdio.h>
+
+#include <board.h>
+#include "board.h"
```
Still waiting for all these includes moved directly to the c-files that need them

vikrant8051 replied:
Done !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_vnd_app/prj.conf:
```diff
@@ -0,0 +1,81 @@
+
+CONFIG_BT_RX_STACK_SIZE=4096
+
+
```
Redundant empty line here (there should ever be at most one)

vikrant8051 replied:
removed !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_vnd_app/src/common.c:
```diff
@@ -0,0 +1,124 @@
+	gpio_add_callback(button_device[0], &button_cb[0]);
+	gpio_pin_enable_callback(button_device[0], SW0_GPIO_PIN);
+
+
```
Redundant empty line (there should only be one)

vikrant8051 replied:
removed !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_vnd_app/src/common.c:
```diff
@@ -0,0 +1,124 @@
+	gpio_add_callback(button_device[1], &button_cb[1]);
+	gpio_pin_enable_callback(button_device[1], SW1_GPIO_PIN);
+
+
```
Redundant empty line (there should only be one)

vikrant8051 replied:
removed !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_vnd_app/src/common.c:
```diff
@@ -0,0 +1,124 @@
+	gpio_add_callback(button_device[2], &button_cb[2]);
+	gpio_pin_enable_callback(button_device[2], SW2_GPIO_PIN);
+
+
```
Redundant empty line (there should only be one)

vikrant8051 replied:
removed !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_vnd_app/src/common.c:
```diff
@@ -0,0 +1,124 @@
+	int power, color;
+
+	power = 100 * (light_state_current.power + 32768)/65535;
+	color = 100 * (light_state_current.color + 32768)/65535;
```
Minor coding style: there should be a space before and after `/`

vikrant8051 commented:
Hi @jhedberg,
Today added Light Lightness Server, Light Lightness Client,  Light CTL Server, Light CTL Client models in Root Element.

It is not yet complete. In coming days, I will try to complete bonding between states & add other required models if necessary. Hope that will be fine for you. 

Thank You !!

vikrant8051 commented:
Hi @sjanc,
Recently, I've added following Lighting Models:
- Light Lightness Server
- Light Lightness Setup Server
- Light CTL Server
- Light CTL Setup Server

Could you please help me to verify that I'm exactly following Bluetooth-SIG Mesh Model Specification ?

Current implementation is still in incomplete state since I've to bind states with each other. Any suggestion or unsolicited details are highly welcome !!

vikrant8051 commented:
@jhedberg @sjanc @carlescufi 
I've implemented state binding between various states. Now as per my assumption, I've completed everything from my side. 

Please review now.

vikrant8051 commented:
@dbkinder I've edited README.txt under the heading of 
1) Overview 
2) Associations of Models with hardware

Plz review it. Other things are as it is. Thank You !!

dbkinder reviewed (approved):
+1 for docs

vikrant8051 commented:
Hi @jhedberg @carlescufi ,
Just for update, now States of Servers are bounded as per Bluetooth SIG Mesh Model Specification v1.0

vikrant8051 commented:
Hi @jhedberg @carlescufi @sjanc 
Just for update, now States of Servers are bounded as per Bluetooth SIG Mesh Model Specification v1.0

carlescufi commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/common.h:
```diff
@@ -0,0 +1,55 @@
+#ifndef _COMMON_H
+#define _COMMON_H
+
+#define  SETB(x, y)   (x |= (1<<y))     /* for o/p */
```
Use Zephyr's `BIT()` macros instead

carlescufi commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/common.c:
```diff
@@ -0,0 +1,147 @@
+	k_work_submit(&button_work);
+}
+
+void gpio_init(void)
```
I think this code and the `button_presed` function could be in `main.c`

carlescufi commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/common.c:
```diff
@@ -0,0 +1,147 @@
+
+#define MINDIFF 2.25e-308
+
+float sqrt(float square)
```
This should probably be added to Zephyr instead, others might want to use it! Not a big deal though

carlescufi commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/common.c:
```diff
@@ -0,0 +1,147 @@
+/* Bluetooth: Mesh Generic OnOff, Generic Level, Lighting & Vendor Models
```
I think if you move the functions here to other files we can avoid having `common.c` and `common.h` which is something that none of  the other apps do

vikrant8051 replied:
If I removed common.c & common.h , then I've to restructure everything. Please allow it to be there if possible. It may be irreverent here but could help others to put  their common things in those files.

vikrant8051 replied:
I've make above changes except removing common.c & common.h which contains some function declarations & pre-processor definitions.

carlescufi replied:
Sounds good, we can always iterate later on future PRs.

vikrant8051 replied:
@carlescufi All checks have passed after latest changes. Meanwhile if possible please request @sjanc to review it to find architecture levels bugs or to scrutinize the things as per Mesh Model specification. 

Thanks for your support !!

vikrant8051 replied:
@carlescufi I've removed common.c as per your recommendation. Hope now you will passed it.

carlescufi reviewed (changes requested):
Hey @vikrant8051, nice job! In general I like it but there's a few things I noted down for you to take a look at. Thank you very much for your contribution.

carlescufi commented:
Thanks @vikrant8051! Approved now.

vikrant8051 commented:
Thank you @carlescufi  !!!!

jhedberg commented:
@vikrant8051 approved by me now as well, however you still need to make the Shippable checks pass. They're failing because of the following type of errors:

```
-:133: ERROR:TRAILING_WHITESPACE: trailing whitespace
#133: FILE: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/README.rst:102:
+The meshctl target for configuration must be the root element's unicast $

-:134: ERROR:TRAILING_WHITESPACE: trailing whitespace
#134: FILE: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/README.rst:103:
+address as it is the only one that has a configuration server model. If $

-:820: WARNING:TYPO_SPELLING: 'vaule' may be misspelled - perhaps 'value'?
#820: FILE: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:263:
+		/* Assume vaule is retrived from Persistence Storage */
```

vikrant8052 committed:
Bluetooth: Mesh: Gen. OnOff, Gen. Level, Lighting & Vendor Models

This is a application demonstrating a Bluetooth mesh node in
which Root element has following models

- Generic OnOff Server
- Generic OnOff Client
- Generic Level Server
- Generic Level Client
- Generic Power OnOff Server
- Generic Power OnOff Setup Server
- Generic Power OnOff Client
- Light Lightness Server
- Light Lightness Setup Server
- Light Lightness Client
- Light CTL Server
- Light CTL Setup Server
- Light CTL Client
- Vendor Model

And Secondary element has following models

- Generic OnOff Server
- Generic OnOff Client
- Generic Level Server
- Generic Level Client
- Light CTL Temperature Server

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
Yes resolving that issue only.

vikrant8051 commented:
@jhedberg Yes resolving that issue only.

vikrant8051 commented:
@jhedberg Thanks for approving it. And thanks for your support.

vikrant8051 commented:
@jhedberg Thanks a lot for reviewing & approving this PR.

I've done best from my side to make it bug free & strictly follow the Mesh Model Specification. Now I'm waiting from developers like me to find the bugs in it, so that I can make it even better :)

vikrant8051 commented:
@jhedberg @carlescufi Thanks a lot for reviewing & approving this PR.

I've done best from my side to make it bug free & strictly follow the Mesh Model Specification. Now I'm waiting for response/feedback from developers like me to find the bugs in it (after it get merged with main branch), so that I can make it even better :).

vikrant8051 commented:
@jhedberg Shippable checks have passed.
